src/wallet/use-cases/register-wallet/register-wallet.spec.ts

```diff
@@ -0,0 +1,100 @@
+import { Test, TestingModule } from '@nestjs/testing';
+import { BadRequestException, NotFoundException } from '@nestjs/common';
+
+import { RegisterWalletUseCase } from '@/wallet/use-cases';
+import { IWalletRepository } from '@/wallet/interfaces';
+import { SaveWalletDTO } from '@/wallet/dto';
+import { AccountType } from '@/shared/constants/enums';
+import { ListUserByIdUseCase } from '@/user/use-cases';
+import { FindBankByIdUseCase } from '@/bank/use-cases';
+import { Wallet } from '@/wallet/infra/entities';
+
+describe('Register Wallet Use Case', () => {
+  let registerWalletUseCase: RegisterWalletUseCase;
+  let walletRepository: IWalletRepository;
+  let listUserByIdUseCase: ListUserByIdUseCase;
+  let findBankByIdUseCase: FindBankByIdUseCase;
+  let createdUser: SaveWalletDTO;
+
+  beforeEach(async () => {
+    const module: TestingModule = await Test.createTestingModule({
+      providers: [
+        RegisterWalletUseCase,
+        {
+          provide: IWalletRepository,
+          useValue: {
+            create: jest.fn(),
+          },
+        },
+        {
+          provide: ListUserByIdUseCase,
+          useValue: {
+            execute: jest.fn(),
+          },
+        },
+        {
+          provide: FindBankByIdUseCase,
+          useValue: {
+            execute: jest.fn(),
+          },
+        },
+      ],
+    }).compile();
+
+    registerWalletUseCase = module.get<RegisterWalletUseCase>(
+      RegisterWalletUseCase,
+    );
+    walletRepository = module.get<IWalletRepository>(IWalletRepository);
+    listUserByIdUseCase = module.get<ListUserByIdUseCase>(ListUserByIdUseCase);
+    findBankByIdUseCase = module.get<FindBankByIdUseCase>(FindBankByIdUseCase);
+
+    createdUser = {
+      user_id: 'user_id',
+      bank_id: 'bank_id',
+      account_type: AccountType.CHECKING_ACCOUNT,
+      description: 'Descrição da carteira',
+    };
+  });
+
+  it('should be able to register a new wallet', async () => {
+    listUserByIdUseCase.execute = jest.fn().mockResolvedValueOnce({});
+    findBankByIdUseCase.execute = jest.fn().mockResolvedValueOnce({});
+    walletRepository.create = jest.fn().mockResolvedValue({} as Wallet);
+
+    const result = await registerWalletUseCase.createWallet(createdUser);
+
+    expect(result).toBeDefined();
+    expect(walletRepository.create).toHaveBeenCalledWith(createdUser);
```

Comment:
Nesses casos acho legal adicionar um expect de quantas vezes aquele método foi chamado pra assegurar que os métodos estão sendo chamados da maneira correta e nos momentos corretos.

Reply:
E lembra que quando você faz esse tipo de teste os seus mocks tem que estar em um beforeEach, caso contrário ele vai contar a chamada de cada um dos testes e não individualmente do que ele está rodando.

Por exemplo, se você tem 5 testes que rodam o método que você está testando ele vai contar 5 chamadas se for em um BeforeAll

no BeforeEach ele gera os mocks de novo toda vez que vai rodar os testes, dessa forma ele só conta a chamada desse teste em especifico

Reply:
Obrigado Rafa e Luiz! Mandaram muito nas dicas, estou aprendendo bastante, ficou perfeito, pois coloquei os mocks no beforeEach.